proctor-common/src/main/java/com/indeed/proctor/common/ForceGroupsOptionsStrings.java

```diff
@@ -38,39 +47,156 @@ public static ForceGroupsOptions parseForceGroupsString(@Nullable final String f
+            if (groupString.isEmpty()) {
                 continue;
             }
 
-            ForceGroupsDefaultMode.fromToken(piece)
+            ForceGroupsDefaultMode.fromToken(groupString)
                     .ifPresent(builder::setDefaultMode);
 
-            int bucketValueStart = piece.length() - 1;
-            for (; bucketValueStart >= 0; bucketValueStart--) {
-                if (!Character.isDigit(piece.charAt(bucketValueStart))) {
-                    break;
+            final Optional<Integer> bucketValueStart = getBucketValueStart(groupString);
+
+            if (bucketValueStart.isPresent()) {
+                //  bucketValueStart should now be the index of the minus sign or the first digit in a run of digits going to the end of the word
+                final String testName = groupString.substring(0, bucketValueStart.get()).trim();
+                final String bucketValueStr = groupString.substring(bucketValueStart.get()).trim();
+                try {
+                    final Integer bucketValue = Integer.valueOf(bucketValueStr);
+                    builder.putForceGroup(testName, bucketValue);
+                    if (bucketAndPayloadValuesStr.length == FORCE_PARAMETER_MAX_SIZE) {
+                        final Payload payloadValue = parseForcePayloadString(bucketAndPayloadValuesStr[FORCE_PARAMETER_PAYLOAD_IDX]);
+                        if (payloadValue != null) {
+                            builder.putForcePayload(testName, payloadValue);
+                        }
+                    }
+                } catch (final NumberFormatException e) {
+                    LOGGER.error("Unable to parse bucket value " + bucketValueStr + " as integer", e);
                 }
             }
-            //  if no name or no value was found, it's not a valid proctor test bucket name
-            if ((bucketValueStart == piece.length() - 1) || (bucketValueStart < 1)) {
-                continue;
-            }
-            //  minus sign can only be at the beginning of a run
-            if (piece.charAt(bucketValueStart) != '-') {
-                bucketValueStart++;
+        }
+        return builder.build();
+    }
+
+    private static Optional<Integer> getBucketValueStart(final String groupString) {
+        int bucketValueStart = groupString.length() - 1;
+        for (; bucketValueStart >= 0; bucketValueStart--) {
+            if (!Character.isDigit(groupString.charAt(bucketValueStart))) {
+                break;
             }
-            //  bucketValueStart should now be the index of the minus sign or the first digit in a run of digits going to the end of the word
-            final String testName = piece.substring(0, bucketValueStart).trim();
-            final String bucketValueStr = piece.substring(bucketValueStart);
+        }
+        //  if no name or no value was found, it's not a valid proctor test bucket name
+        if ((bucketValueStart == groupString.length() - 1) || (bucketValueStart < 1)) {
```

Comment:
will it be same and easier to understand if we do 

if (bucketValueStr.isEmpty() || testName.isEmpty())

?

Reply:
This is looking to see within the bucketValueStr if there is a digit that will represent the Bucket value, so replacing that would not be equivalent.

Reply:
bucketValueStr will only be created if bucketValueStart is a valid index. In your question the bucketValueStart is not a valid index and thus will return an Optional.Empty which will skip attempting to add this value as a force Group. 

The cases for an invalid index would be a test bucket with no digit (e.g. testbucket), a value that is only digits e.g. (112241231), an empty string.

Reply:
Yes and thanks for the explanation.

Given 
final String testName = piece.substring(0, bucketValueStart).trim();
final String bucketValueStr = piece.substring(bucketValueStart);

And the following condition
if ((bucketValueStart == groupString.length() - 1) || (bucketValueStart < 1))

When bucketValueStart == groupString.length() - 1, bucketValueStr = empty string, testName = full string
When bucketValueStart < 1(actually in this case it bucketValueStart will be equal to 0), bucketValueStr = full string, testName = empty string.

In my understanding 
if ((bucketValueStart == groupString.length() - 1) || (bucketValueStart < 1)) is harder to read
compared to 
if (testName.isEmpty() || bucketValueStr.isEmpty()) because "testName.isEmpty()" means no test name and "bucketValueStr.isEmpty()" means no digit is found.
This is just a personal opinion and I dont have a strong preference either way.


What do you think?

Reply:
I would prefer to leave as is. This code was existing and just moved into a separate function to make parseForceGroupsString() more readable.